libp2p/io/msgio.py

```diff
@@ -5,80 +5,85 @@
 
 NOTE: currently missing the capability to indicate lengths by "varint" method.
 """
-# TODO unify w/ https://github.com/libp2p/py-libp2p/blob/1aed52856f56a4b791696bbcbac31b5f9c2e88c9/libp2p/utils.py#L85-L99  # noqa: E501
-from typing import Optional, cast
+from abc import abstractmethod
 
-from libp2p.io.abc import Closer, ReadCloser, Reader, ReadWriteCloser, WriteCloser
+from libp2p.io.abc import MsgReadWriteCloser, Reader, ReadWriteCloser
 from libp2p.io.utils import read_exactly
+from libp2p.utils import decode_uvarint_from_stream, encode_varint_prefixed
+
+from .exceptions import MessageTooLarge
 
-SIZE_LEN_BYTES = 4
 BYTE_ORDER = "big"
 
 
-async def read_length(reader: Reader) -> int:
-    length_bytes = await read_exactly(reader, SIZE_LEN_BYTES)
+async def read_length(reader: Reader, size_len_bytes: int) -> int:
+    length_bytes = await read_exactly(reader, size_len_bytes)
     return int.from_bytes(length_bytes, byteorder=BYTE_ORDER)
 
 
-def encode_msg_with_length(msg_bytes: bytes) -> bytes:
-    len_prefix = len(msg_bytes).to_bytes(SIZE_LEN_BYTES, "big")
+def encode_msg_with_length(msg_bytes: bytes, size_len_bytes: int) -> bytes:
+    try:
+        len_prefix = len(msg_bytes).to_bytes(size_len_bytes, byteorder=BYTE_ORDER)
+    except OverflowError:
+        raise ValueError(
+            "msg_bytes is too large for `size_len_bytes` bytes length: "
+            f"msg_bytes={msg_bytes!r}, size_len_bytes={size_len_bytes}"
+        )
     return len_prefix + msg_bytes
 
 
-class MsgIOWriter(WriteCloser):
-    write_closer: WriteCloser
-
-    def __init__(self, write_closer: WriteCloser) -> None:
-        self.write_closer = write_closer
-
-    async def write(self, data: bytes) -> int:
-        await self.write_msg(data)
-        return len(data)
+class BaseMsgReadWriter(MsgReadWriteCloser):
+    read_write_closer: ReadWriteCloser
+    size_len_bytes: int
 
-    async def write_msg(self, msg: bytes) -> None:
-        data = encode_msg_with_length(msg)
-        await self.write_closer.write(data)
+    def __init__(self, read_write_closer: ReadWriteCloser) -> None:
+        self.read_write_closer = read_write_closer
 
-    async def close(self) -> None:
-        await self.write_closer.close()
+    async def read_msg(self) -> bytes:
+        length = await self.next_msg_len()
+        return await read_exactly(self.read_write_closer, length)
 
+    @abstractmethod
+    async def next_msg_len(self) -> int:
+        ...
 
-class MsgIOReader(ReadCloser):
-    read_closer: ReadCloser
-    next_length: Optional[int]
+    @abstractmethod
+    def encode_msg(self, msg: bytes) -> bytes:
```

Comment:
should this be public?  Is it used as an external API or only called internally?

Reply:
It's only called internally, but I expect subclasses to implement it to make `write_msg ` work. So as `next_msg_len` to make `read_msg` work.